### PR TITLE
Allows to check equality for objects with not only primitive types and display span information for ScriptRuntimeException

### PR DESCRIPTION
**Pull request by tatarincev**
**Objects equality**
support the follow syntax {{ obj1 == obj2 }}

**Display span information for ScriptRuntimeException**

Before
image

After
image

Excuse me for change formatting, I hope you will fix it during merging

### CODE REVIEW

File: src/Scriban/Syntax/ScriptRuntimeException.cs [hunk 1]
```diff
@@ -11,17 +11,16 @@ namespace Scriban.Syntax
...
     public class ScriptRuntimeException : Exception
     {
-        public ScriptRuntimeException(SourceSpan span, string message) : base(message)
+        public ScriptRuntimeException(SourceSpan span, string message) : this(span, message, null)
         {
-            Span = span;
         }
 
-        public ScriptRuntimeException(SourceSpan span, string message, Exception innerException) : base(message, innerException)
+        public ScriptRuntimeException(SourceSpan span, string message, Exception innerException) : base(new LogMessage(ParserMessageType.Error, span, message).ToString(), innerException)
```
[the remark this answers is not in the document]
Comment (xoofx) [reply src/Scriban/Syntax/ScriptRuntimeException.cs]: The ToString() of this object is already doing this, I'm confused why you want to do this here (haven't look deep into the code and I don't recall this part, I apologize)

[the remark this answers is not in the document]
Comment (tatarincev) [reply src/Scriban/Syntax/ScriptRuntimeException.cs]: Asp. Net core doesn't call exception.ToString for display errors details. Seems it uses exception.Message instead

[the remark this answers is not in the document]
Comment (xoofx) [reply src/Scriban/Syntax/ScriptRuntimeException.cs]: Ok but we should not change this. Can't you try/catch and rethrow with a message if it is your usacase?

[the remark this answers is not in the document]
Comment (tatarincev) [reply src/Scriban/Syntax/ScriptRuntimeException.cs]: Ok, I'll do it on my application level. I've been removed this code from PR.
However, I still thinking that it is not  the right way to override  Exception.ToString() for this purposes.

[the remark this answers is not in the document]
Comment (tatarincev) [reply src/Scriban/Syntax/ScriptRuntimeException.cs]: done

File: src/Scriban/Syntax/ScriptRuntimeException.cs [hunk 2]
```diff
@@ -11,21 +11,31 @@ namespace Scriban.Syntax
 {
     public class ScriptRuntimeException : Exception
     {
+        private string _message;
```
Comment (xoofx) [inline src/Scriban/Syntax/ScriptRuntimeException.cs]: No need to add _message as this is accessible via `base.Message`

File: src/Scriban/Syntax/ScriptRuntimeException.cs [hunk 3]
```diff
@@ -11,21 +11,31 @@ namespace Scriban.Syntax
...
+        {
+            get
+            {
+                return ToString();
+            }
+        }
         public override string ToString()
         {
-            return new LogMessage(ParserMessageType.Error, Span, Message).ToString();
+            return new LogMessage(ParserMessageType.Error, Span, _message).ToString();
```
Comment (xoofx) [inline src/Scriban/Syntax/ScriptRuntimeException.cs]: Could you remove the ToString() method entirely and move this to the Message property directly? Than use `base.Message` instead

Comment (tatarincev) [reply src/Scriban/Syntax/ScriptRuntimeException.cs]: Done

**Comment by xoofx** [comment]
Thanks!

At least, the last remaining issue is:

> Also you will need to run AsyncCodeGen as they need to be kept in sync (sorry, I have forgotten to update the doc regarding this)

And add a test for this feature (an object that is passed but not a primitive for the compare equal test)

**Comment by tatarincev** [comment]
I ran Scriban.CodeGen.exe, but it looks like didn't generate anything.
image
What about tests, I added some new tests into `111-binary-compare`. but alos  unfortunately can't run it by unknown reasons
image

**Comment by xoofx** [comment]
`AsyncCodeGen` not CodeGen. 😉  CodeGen is when you change a function.

File: src/Scriban.Tests/TestFiles/100-expressions/111-binary-compare.txt [hunk 4]
```diff
@@ -47,3 +47,9 @@ a == empty: {{
     a == empty
 }}
 b == empty: {{b == empty}}
+
```
Comment (xoofx) [inline src/Scriban.Tests/TestFiles/100-expressions/111-binary-compare.txt]: Remove this extra line. This is what is failing the roundtrip test

Comment (tatarincev) [reply src/Scriban.Tests/TestFiles/100-expressions/111-binary-compare.txt]: I've been removed the trail spaces in test files, but unfortunately I have no luck with run it on my machine. NUnit cant discover this tests.  I don't have any experience with NUnit so I hope for you help.
`
 No test matches the given testcase filter `FullyQualifiedName=Scriban.Tests.TestParser.A100_expressions` in C:\Projects\VirtoCommerce\scriban\src\Scriban.Tests\bin\Debug\netcoreapp2.1\Scriban.Tests.dll
`

And `AsyncCodeGen ` generate invalid code. See screenshot in message above.

**Comment by xoofx** [comment]
Thanks, this is merged.

I fixed the space issue (in both files by commit 0baf725fc8b000331f9a46f6f13c19ff451bebeb

AsyncCodeGen was actually not necessarily to run, but I tried to fix the issue for running it by forcing `net35` in commit b8162dae6ae9233ae4c8aebf8d175194ec7a4274

**Comment by tatarincev** [comment]
Please excuse me I for being stupid, a lot of  work on our  the project and not have much time to better understand and deep into your project. In the future PR I'll try to do all the best at first commit.  
I'm working on DotLiquid replacement to Scriban due to very poor performance and extensibility. I think you will be pleasure to know what migration to scriban render engine increased the application performance u to 3 times! Big thanks for you work!
image

**Comment by xoofx** [comment]
> Please excuse me I for being stupid, a lot of work on our the project and not have much time to better understand and deep into your project. 

No worries, but it is good for your project to know how to navigate into Scriban's codebase in order to fix bugs or bring improvements, as I might not be able to always respond promptly or have time/interest to implement a particular feature.

> scriban render engine increased the application performance up to 3 times! Big thanks for you work!

Super nice, thanks for the feedback! 😉

**Pull request merged**